File: src/main/java/gg/mineral/practice/inventory/menus/ConfigureValueMenu.java
package gg.mineral.practice.inventory.menus;

import java.util.function.Consumer;
import gg.mineral.practice.entity.Profile;
import gg.mineral.practice.inventory.AnvilMenu;
import gg.mineral.practice.inventory.ClickCancelled;
import gg.mineral.practice.inventory.Menu;
import gg.mineral.practice.util.items.ItemStacks;
import gg.mineral.practice.util.messages.impl.ErrorMessages;
import lombok.RequiredArgsConstructor;

@ClickCancelled(true)
@RequiredArgsConstructor(access = lombok.AccessLevel.PRIVATE)
public class ConfigureValueMenu<T> extends AnvilMenu {
    private final Menu menu;
    private final Consumer<T> value;
    private final Class<T> type;

    @Override
    @SuppressWarnings("unchecked")
    public void update() {

        setSlot(1, ItemStacks.APPLY, interaction -> {
            Profile p = interaction.getProfile();

            String text = getText();

            if (text == null) {
                p.message(ErrorMessages.INVALID_NUMBER);
                return;
            }

            try {
                if (type.equals(Double.class) || type.equals(double.class))
                    value.accept((T) Double.valueOf(text.replace(" ", "")));
                else if (type.equals(Float.class) || type.equals(float.class))
                    value.accept((T) Float.valueOf(text.replace(" ", "")));
                else if (type.equals(Integer.class) || type.equals(int.class))
                    value.accept((T) Integer.valueOf(text.replace(" ", "")));
                else if (type.equals(Long.class) || type.equals(long.class))
                    value.accept((T) Long.valueOf(text.replace(" ", "")));
                else if (type.equals(Short.class) || type.equals(short.class))
                    value.accept((T) Short.valueOf(text.replace(" ", "")));
                else if (type.equals(Byte.class) || type.equals(byte.class))
                    value.accept((T) Byte.valueOf(text.replace(" ", "")));
                else
                    throw new IllegalArgumentException("Unsupported type: " + type.getTypeName());
            } catch (NumberFormatException e) {
                p.message(ErrorMessages.INVALID_NUMBER);
                return;
            }
            p.openMenu(menu);
        });

        setSlot(0, ItemStacks.CANCEL, interaction -> interaction.getProfile().openMenu(menu));
    }

    public static <T> ConfigureValueMenu<T> of(Menu menu, Consumer<T> value, Class<T> type) {
        return new ConfigureValueMenu<>(menu, value, type);
    }
}

File: src/main/kotlin/gg/mineral/practice/inventory/menus/ConfigureValueMenu.kt
package gg.mineral.practice.inventory.menus

import gg.mineral.practice.inventory.AnvilMenu
import gg.mineral.practice.inventory.ClickCancelled
import gg.mineral.practice.inventory.Interaction
import gg.mineral.practice.inventory.Menu
import gg.mineral.practice.util.items.ItemStacks
import gg.mineral.practice.util.messages.impl.ErrorMessages
import java.util.function.Consumer
import kotlin.reflect.KClass

@ClickCancelled(true)
class ConfigureValueMenu<T : Number> private constructor(
    private val menu: Menu,
    private val value: Consumer<T>,
    private val kclass: KClass<T>
) : AnvilMenu() {

    override fun update() {
        setSlot(1, ItemStacks.APPLY) { interaction: Interaction ->
            val profile = interaction.profile
            val inputText = text?.trim() ?: run {
                profile.message(ErrorMessages.INVALID_NUMBER)
                return@setSlot
            }

            // Try to parse inputText as whichever Kotlin numeric type we expect.
            val parsed: Number = try {
                when (kclass) {
                    Int::class -> inputText.toInt()
                    Double::class -> inputText.toDouble()
                    Float::class -> inputText.toFloat()
                    Long::class -> inputText.toLong()
                    Short::class -> inputText.toShort()
                    Byte::class -> inputText.toByte()

                    // If you only want to handle these six Kotlin numeric types, throw here:
                    else -> throw IllegalArgumentException(
                        "Unsupported numeric type: ${kclass.simpleName}"
                    )
                }
            } catch (e: NumberFormatException) {
                profile.message(ErrorMessages.INVALID_NUMBER)
                return@setSlot
            }

            // Unsafe cast but fine if the above 'when' matches the correct type:
            @Suppress("UNCHECKED_CAST")
            value.accept(parsed as T)

            profile.openMenu(menu)
        }

        setSlot(0, ItemStacks.CANCEL) { it.profile.openMenu(menu) }
    }

    companion object {
        fun <T : Number> of(
            menu: Menu,
            value: Consumer<T>,
            kclass: KClass<T>
        ): ConfigureValueMenu<T> {
            return ConfigureValueMenu(menu, value, kclass)
        }
    }
}

